docs/_includes/Expressions.md

```diff
@@ -416,21 +416,55 @@ or a collection of object references.
 
 TO BE WRITTEN -- allocated predicate
 
-## Old Expressions {#sec-old-expression}
+## Old and Old@ Expressions {#sec-old-expression}
 
-````grammar
+````
 OldExpression_ = "old" [ "@" ident ] "(" Expression(allowLemma: true, allowLambda: true) ")"
 ````
 
-An _old expression_ is used in postconditions. `old(e)` evaluates to
-the value expression `e` had on entry to the current method.
-Note that **old** only affects heap dereferences, like `o.f` and `a[i]`.
-In particular, **old** has no effect on the value returned for local
-variables or out-parameters.
+An _old expression_ is used in postconditions or in the body of a method
+or in the body or specification of any two-state function or two-state lemma;
+an _old_ expression with a label is used only in the body of a method at a point
+where the label dominates its use in this expression.
+
+`old(e)` evaluates
+the argument using the value of the heap on entry to the method;
+`old@ident(e)` evaluates the argument using the value of the heap at the
+given statement label.
+
+Note that **old** and **old@** only affect heap dereferences,
+like `o.f` and `a[i]`.
+In particular, neither form has any effect on the value returned for local
+variables or out-parameters or ghost values (as they are not on the heap).[^Old]
```

Comment:
```suggestion
variables or out-parameters (as they are not on the heap).[^Old]
```

The ghost-ness of something does not play into the semantics of `old`. In particular, if `x` is a local variable, then `old(x)` means the same thing regardless of if `x` is ghost or compiled. Also, if `f` is a field (or a `class` or `trait`), then `old(x.f)` means the same thing regardless of if `f` is ghost or compiled.

Reply:
OK